### PR TITLE
Added a tool to split the packages from our build system into multiple directories based on a CSV file

### PR DESCRIPTION
**Pull request by javiercn**

### CODE REVIEW

**Comment by javiercn** [comment]
@pranavkm @muratg

**Comment by pranavkm** [comment]
:shipit:

**Comment by muratg** [comment]
:shipit:

File: src/SplitPackages/Program.cs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            public bool FoundInFolder { get; set; }
+
+            public bool Equals(PackageItem other)
+            {
+                return string.Equals(OriginPath, other?.OriginPath);
+            }
+
+            public override int GetHashCode()
+            {
+                return OriginPath?.GetHashCode() * 7 ?? 1;
```
Comment (Eilon) [inline src/SplitPackages/Program.cs]: As a general rule, avoid being fancy here. Just return the string's hash code and be done.

File: src/SplitPackages/Program.cs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                package.DestinationPath = path;
+            }
+        }
+
+        private static void CopyPackages(IList<PackageItem> expandedPackages, bool whatif)
+        {
+            foreach (var package in expandedPackages)
+            {
+                WriteInformational($@"Copying package {package.Name} from
+{package.OriginPath} to
```
Comment (Eilon) [inline src/SplitPackages/Program.cs]: To make this more readable, indent the sub-lines.

File: src/SplitPackages/Program.cs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        {
+            if (args.Contains("help", StringComparer.OrdinalIgnoreCase) ||
+                args.Length < 3 ||
+                args.Length > 4)
+            {
+                PrintUsage();
+                return;
+            }
+
+            bool whatif = args.Length == 4 && args[3].Equals("whatif", StringComparison.OrdinalIgnoreCase);
```
Comment (moozzyk) [inline src/SplitPackages/Program.cs]: var?

File: src/SplitPackages/Program.cs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+        private static Arguments NormalizeArguments(string[] args)
+        {
+            var sourcePackages = Path.GetFullPath(args[0]);
+            var csvFile = Path.GetFullPath(args[1]);
+            var destinationFolder = Path.GetFullPath(args[2]);
+
+            if (!Directory.Exists(sourcePackages))
+            {
+                var msg = $@"Source packages folder does not exist or is not a folder
```
Comment (moozzyk) [inline src/SplitPackages/Program.cs]: is `@` required at all? Also why do you need `msg`?  Can't you just inline the message?

File: src/SplitPackages/Program.cs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace SplitPackages
+{
+    class Program
+    {
+        static void Main(string[] args)
```
Comment (Eilon) [inline src/SplitPackages/Program.cs]: Return `int` and return non-zero value for failures.

File: src/SplitPackages/Program.cs (hunk 3 above)
Comment (Eilon) [inline src/SplitPackages/Program.cs]: This option isn't documented in the help.

File: src/SplitPackages/Program.cs [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+            Console.WriteLine($@"There was an error executing the command:
+{error}");
+            Console.ResetColor();
+            PrintUsage();
+        }
+
+        private static void PrintUsage()
+        {
+            var text = @"
+SplitPackages [help] [source packages] [csv path] [destination folder]
```
Comment (Eilon) [inline src/SplitPackages/Program.cs]: This help doesn't match the actual usage. Instead I recommend showing the 3-4 examples of usage:

```
SplitPackages help
SplitPackages whatif X Y Z
SplitPackages X Y Z
```

And then document what each parameter uses. This avoids ambiguity in how options can be combined.

File: src/SplitPackages/Program.cs [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+                OriginPath = p,
+                Name = Path.GetFileName(p),
+                FoundInFolder = true
+            }).ToList();
+        }
+
+        private static IList<PackageItem> GetPackagesFromCsvFile(string csvFile)
+        {
+            var lines = File.ReadAllLines(csvFile);
+            var parsedLines = lines.Select(l => l.Split(','));
```
Comment (moozzyk) [inline src/SplitPackages/Program.cs]: CSV has some escaping rules - not sure if they need to be taken into account or not.

File: src/SplitPackages/Program.cs [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+                DestinationFolderName = parsed[1].Trim(),
+                FoundInCsv = true
+            }).ToList();
+        }
+
+        private static IList<PackageItem> ExpandPackages(IList<PackageItem> packagesFromCsv, string sourceFolder)
+        {
+            var allPackages = new HashSet<PackageItem>();
+
+            for (int i = 0; i < packagesFromCsv.Count; i++)
```
Comment (moozzyk) [inline src/SplitPackages/Program.cs]: nit: var

File: src/SplitPackages/Program.cs [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+using System.IO;
+using System.Linq;
+
+namespace SplitPackages
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            if (args.Contains("help", StringComparer.OrdinalIgnoreCase) ||
```
Comment (victorhurdugaci) [inline src/SplitPackages/Program.cs]: Why don't use use the command line parsing utils?

File: src/SplitPackages/Program.cs [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+                if (!whatif)
+                {
+                    File.Copy(package.OriginPath, package.DestinationPath, overwrite: true);
+                }
+            }
+        }
+
+        private static void WriteInformational(string msg)
+        {
+            Console.ForegroundColor = ConsoleColor.Cyan;
```
Comment (victorhurdugaci) [inline src/SplitPackages/Program.cs]: Why not use our logger?

File: src/SplitPackages/Program.cs [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+{csvFile}";
+                throw new InvalidOperationException(msg);
+            }
+
+            if (!Directory.Exists(destinationFolder))
+            {
+                var msg = $@"Destination folder does not exist, creating folder at:
+{destinationFolder}";
+                WriteInformational(msg);
+                Directory.CreateDirectory(destinationFolder);
```
Comment (victorhurdugaci) [inline src/SplitPackages/Program.cs]: It's a little strange that `NormalizeArguments` creates a folder...

File: src/SplitPackages/Program.cs [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+                if (expandedPackages.Count > 0)
+                {
+                    packagesFromCsv[i].FoundInFolder = true;
+                }
+            }
+
+            return allPackages.ToList();
+        }
+
+        private static void EnsureNoExtraSourcePackagesFound(IList<PackageItem> source, IList<PackageItem> expanded)
```
Comment (victorhurdugaci) [inline src/SplitPackages/Program.cs]: Nit: you could do this in 1 line with Linq's `Except` :)

File: src/SplitPackages/Program.cs (hunk 7 above)
Comment (Eilon) [inline src/SplitPackages/Program.cs]: Yeah this clearly isn't a proper CSV parser, but in our case the inputs are highly constrained.

Comment (moozzyk) [inline src/SplitPackages/Program.cs]: @Eilon :+1:

File: src/SplitPackages/Program.cs (hunk 1 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: cleaned up

File: src/SplitPackages/Program.cs (hunk 2 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: done in https://github.com/aspnet/DnxTools/pull/17

File: src/SplitPackages/Program.cs (hunk 3 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: this line is removed, I've updated the tool to use our command line utils in https://github.com/aspnet/DnxTools/pull/17

File: src/SplitPackages/Program.cs (hunk 4 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: @ is required as it's a multiline string literal, inlining makes everything uglier

File: src/SplitPackages/Program.cs (hunk 5 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: done in https://github.com/aspnet/DnxTools/pull/17

File: src/SplitPackages/Program.cs (hunk 6 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: This uses the standard help from our command line utilities

File: src/SplitPackages/Program.cs (hunk 7 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: Yeah, YAGNI

File: src/SplitPackages/Program.cs (hunk 9 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: changed

File: src/SplitPackages/Program.cs (hunk 10 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: changed

File: src/SplitPackages/Program.cs (hunk 12 above)
Comment (javiercn) [inline src/SplitPackages/Program.cs]: Still need to go through the list of expanded packages to mark them as found in source, so it's fine like it is right now

**Pull request merged**